CMakeLists.txt

```diff
@@ -1,12 +1,205 @@
+# Toolchain must come before first project
+set(CMAKE_TOOLCHAIN_FILE ${CMAKE_SOURCE_DIR}/arm-none-eabi-gcc.cmake
+  CACHE FILEPATH "CMake ARM GCC toolchain file" FORCE
+)
+
 cmake_minimum_required(VERSION 3.15)
-project(stm32_spinner_project-2)
+project(stm32_spinner_project)
+enable_language(C ASM)
 set(CMAKE_CXX_STANDARD 11)
 set(CMAKE_CXX_STANDARD_REQUIRED true)
 set(CMAKE_C_STANDARD 11)
 set(CMAKE_C_STANDARD_REQUIRED true)
+
+# Build type
+set(EXECUTABLE_OUTPUT_PATH  ${CMAKE_SOURCE_DIR}/build/${CMAKE_BUILD_TYPE})
+
+set(CMAKE_EXECUTABLE_SUFFIX ".elf")
+
+# set location of project
+set(PRJ_DIR ${CMAKE_SOURCE_DIR}/lib/STM32Cube_FW_F4_V1.25.1)
+
+# set startup file
+set(STARTUP ${PRJ_DIR}/Drivers/CMSIS/Device/ST/STM32F4xx/Source/Templates/gcc/startup_stm32f407xx.s)
+message (STATUS "[DBG]: ${STARTUP}")
+
+# set system file for Controller
+set(SYSTEM ${PRJ_DIR}/Projects/STM32F4-Discovery/Examples/GPIO/GPIO_EXTI/Src/system_stm32f4xx.c)
+
+# set paths
+set(HAL_PATH ${PRJ_DIR}/Drivers/STM32F4xx_HAL_Driver)
+set(CMSIS_PATH ${PRJ_DIR}/Drivers/CMSIS)
+set(BSP_PATH ${PRJ_DIR}/Drivers/BSP)
     
-set(EXTERNAL_PROJECT_GENERATOR "MinGW Makefiles")
-set(CMAKE_SH_NOT_FOUND "-DCMAKE_SH=CMAKE_SH-NOTFOUND")
-set(EXTERNAL_PROJECT_COMPILER "-DCMAKE_C_COMPILER=gcc")
 
-add_executable(stm32_spinner_project-2 Core/Src/main.c)
+#add_executable(stm32_spinner_project-2 Core/Src/main.c)
+
+# set CMSIS files needed for the project
+set(CMSIS_SRCS  
+  ${CMSIS_PATH}/Device/ST/STM32F4xx/Include/stm32f4xx.h
+  ${CMSIS_PATH}/Device/ST/STM32F4xx/Include/stm32f407xx.h
+  ${CMSIS_PATH}/Core/Include/cmsis_gcc.h
+  ${CMSIS_PATH}/Core/Include/cmsis_armcc.h
+  ${CMSIS_PATH}/Core/Include/cmsis_compiler.h
+  ${CMSIS_PATH}/Core/Include/cmsis_version.h
+  ${CMSIS_PATH}/Core/Include/core_cm4.h
+  ${CMSIS_PATH}/Core/Include/mpu_armv7.h
+)
+
+# set HAL files needed for the project
+set(HAL_SRCS
+  ${HAL_PATH}/Src/stm32f4xx_hal_pwr.c
+  ${HAL_PATH}/Inc/stm32f4xx_hal_pwr.h
+  ${HAL_PATH}/Src/stm32f4xx_hal_cortex.c
+  ${HAL_PATH}/Inc/stm32f4xx_hal_cortex.h
+  ${HAL_PATH}/Src/stm32f4xx_hal_gpio.c
+  ${HAL_PATH}/Inc/stm32f4xx_hal_gpio.h
+  ${HAL_PATH}/Src/stm32f4xx_hal_flash.c
+  ${HAL_PATH}/Inc/stm32f4xx_hal_flash.h
+  ${HAL_PATH}/Src/stm32f4xx_hal_spi.c
+  ${HAL_PATH}/Inc/stm32f4xx_hal_spi.h
+  ${HAL_PATH}/Src/stm32f4xx_hal_rcc.c
+  ${HAL_PATH}/Inc/stm32f4xx_hal_rcc.h
+  ${HAL_PATH}/Src/stm32f4xx_hal_flash_ex.c
+  ${HAL_PATH}/Inc/stm32f4xx_hal_flash_ex.h
+  ${HAL_PATH}/Src/stm32f4xx_hal_rcc_ex.c
+  ${HAL_PATH}/Inc/stm32f4xx_hal_rcc_ex.h
+  ${HAL_PATH}/Src/stm32f4xx_hal_i2c.c
+  ${HAL_PATH}/Inc/stm32f4xx_hal_i2c.h
```

Comment:
You don't need to add include files. just use 
set(INCLUDE_DIRS ${INCLUDE_DIRS} ${....)
or/and
add_subdirectory(...)

Reply:
done